charts/external-dns/values.yaml

```diff
@@ -26,8 +26,12 @@ rbac:
   create: true
   additionalPermissions: {}
 
+# Annotations to add to the Deployment
+annotations: {}
```

Comment:
I think this might be better named as `deploymentAnnotations` unless you've got a strong idiomatic precedent?

Reply:
I have added an example use case in my PR description above. And unfortunately, some tools only look for deploymentAnnotations and not PodAnnotations.

Reply:
@beastob have you looked at adding a [kiwigrid/k8s-sidecar](https://github.com/kiwigrid/k8s-sidecar)? This is how Grafana handles configmap reloads and it seems to be a better architecture with a single responsibility container in a pod rather than a central service? If this worked it'd be easy to add configuration to enable extra containers in the pod and documentation to show how it could be used.

Reply:
Suppose [https://github.com/kiwigrid/k8s-sidecar](https://github.com/kiwigrid/k8s-sidecar) works in a way that's similar to [Spring-boot's ConfigMap/Secret watcher](https://docs.spring.io/spring-cloud-kubernetes/docs/current/reference/html/#monitoring-configmaps-and-secrets). My answer would be yes and no. It's because watchers like kiwigrid/k8s-sidecar would restart(I supposed?) the application container upon config changes, and if the deployment has replica>1, all pods/containers will all restart at the same time and causes short period of down time. However, for this project's I think adding an `extraContainers` field is a good idea, for those who would like to add sidecars.

staketer-reloader on the other hand can allow a RollingUpgrade effect to restart all deployment pods. Nevertheless, I believe allowing a common field `annotations` for Deployment shouldn't be a bad idea?

Reply:
The preferred use of kiwigrid/k8s-sidecar is to hit a webhook to trigger a config reload but I think it can manually re-start a pod if a webhook isn't available.

Does ExternalDNS support running multiple replicas? I'm pretty sure you should only have a single "active" replica and I don't think having a non-leader replica would add a benefit to the ExternalDNS workflow even if possible? Even then assuming multiple pods, if configuration has changed I doubt you'd want some using old config? Finally I can't see ExternalDNS as a tier 1 service, I can't see any use case where unavailable for a restart cycle is a problem?

> Nevertheless, I believe allowing a common field annotations for Deployment shouldn't be a bad idea?

It's not very idiomatic and promotes the tight coupling of controller to pod which should be avoided as much as possible.

Reply:
> Does ExternalDNS support running multiple replicas? I'm pretty sure you should only have a single "active" replica and I don't think having a non-leader replica would add a benefit to the ExternalDNS workflow even if possible? Even then assuming multiple pods, if configuration has changed I doubt you'd want some using old config? Finally I can't see ExternalDNS as a tier 1 service, I can't see any use case where unavailable for a restart cycle is a problem?

I agree that ExternalDNS in particular does not need RollingUpgrade. It's just that using a centralised service like stakater-reloader we don't have to worry if a particular application out of dozens can support hot config reload or not, and we don't need config extra sidecar individually to achieve config reload. Only annotations would be needed.

And in GitOps approach of managing the kubernates cluster, we would want as little manual operations as possible. For example, changing some config in Git and then pushing it. Then the applications should reload the configs without further human interactions.

Reply:
I can agree to disagree on this approach.

So just the CLA error to fix and then someone can enable the CI to run, the code itself looks good to me.

Reply:
I do see your concerns, but it's the lesser of two evils in our case. I will try to figure out the CLA problem later, and thanks for the discussion.